packages/amesos2/src/Amesos2_Tacho_def.hpp

```diff
@@ -0,0 +1,371 @@
+  Teuchos::RCP<const Matrix> A,
+  Teuchos::RCP<Vector>       X,
+  Teuchos::RCP<const Vector> B )
+  : SolverCore<Amesos2::TachoSolver,Matrix,Vector>(A, X, B)
+  , nzvals_()                   // initialize to empty arrays
+  , colind_()
+  , rowptr_()
+{
+}
+
+
+template <class Matrix, class Vector>
+TachoSolver<Matrix,Vector>::~TachoSolver( )
+{
+}
+
+template <class Matrix, class Vector>
+std::string
+TachoSolver<Matrix,Vector>::description() const
+{
+  std::ostringstream oss;
+  oss << "Tacho solver interface";
+  return oss.str();
+}
+
+template<class Matrix, class Vector>
+int
+TachoSolver<Matrix,Vector>::preOrdering_impl()
+{
+  return(0);
+}
+
+template <class Matrix, class Vector>
+int
+TachoSolver<Matrix,Vector>::symbolicFactorization_impl()
+{
+  int status = 0;
+
+  if ( this->root_ ) {
+    size_type_array row_ptr;
+    ordinal_type_array cols;
+
+#ifndef HAVE_TEUCHOS_COMPLEX
+    if(single_process_optim_check()) {
+      // in the optimized case we read the values directly from the matrix
+      // without converting through the Teuchos::Array setup. Note that in
+      // this case nzvals_, colind_, and rowptr_ are never set in loadA_impl.
+      // For HAVE_TEUCHOS_COMPLEX is_optimized_case() return false because we
+      // need nzvals_.
+      auto sp_rowptr = this->matrixA_->returnRowPtr();
+        TEUCHOS_TEST_FOR_EXCEPTION(sp_rowptr == nullptr,
+          std::runtime_error, "Amesos2 Runtime Error: sp_rowptr returned null");
+      auto sp_colind = this->matrixA_->returnColInd();
+        TEUCHOS_TEST_FOR_EXCEPTION(sp_colind == nullptr,
+          std::runtime_error, "Amesos2 Runtime Error: sp_colind returned null");
+
+      // Tacho uses size_type size_t which matches Tpetra but not Epetra (int)
+      // So we need a converter
+      row_ptr = size_type_array("r", this->globalNumRows_ + 1);
+      for(global_size_type n = 0; n < this->globalNumRows_ + 1; ++n) {
+        row_ptr(n) = static_cast<size_type>(sp_rowptr[n]);
+      }
+
+      // Now convert the cols - now on an earlier version I thought this was
+      // working as a straight conversion but then some things have changed
+      // and it seems we will need to convert. TODO: Assess this conversion.
+      // Can we do this more optimally and still work for all Epetra/Tpetra
+      // cases.
+      cols = ordinal_type_array("c", this->globalNumNonZeros_);
+      for(global_size_type n = 0; n < this->globalNumNonZeros_; ++n) {
+        cols(n) = static_cast<ordinal_type>(sp_colind[n]);
+      }
+
+      // TODO - For Tpetra, we could have a direct view like this...
+      // Can we explot this here?
+      // row_ptr = size_type_array(sp_rowptr, this->globalNumRows_ + 1);
+    }
+    else
+#endif
+    {
+      // Non optimized case used the arrays set up in loadA_impl
+      row_ptr = size_type_array(this->rowptr_.getRawPtr(), this->globalNumRows_ + 1);
+      cols = ordinal_type_array(this->colind_.getRawPtr(), this->globalNumNonZeros_);
+    }
+
```

Comment:
row_ptr and cols are type defined from Tacho. Tacho does not have template parameters for local/global ordinals. 

How do I know that this->rowptr_ has the same type while Tpetra can instanciate different combinations of local/global ordinals ?

Reply:
rowptr_ is setup to be type Tacho::Experimental::size_type as a Teuchos::Array and this is filled by Util::get_crs_helper in the loadA_impl so I think this handles all the type needs. I had some optimizations for a single process case (similar to Basker) which Siva and I discussed briefly but I wasn't sure if this can benefit Tacho right now unless we changes Tacho to handle different types coming in. Something I will need to get further clarification on. I have temporarily left that commented out for discussion.